vehicle/remote_control_process.py

```diff
@@ -246,1086 +306,939 @@ def close_motor_socket(self):
-                debug_time = time.time()
-                if self.latest_gps_sample is not None:
-                    self.last_good_gps_sample = self.latest_gps_sample
-                else:
-                    # Occasional bad samples are fine. A very old sample will
-                    # get flagged in the final checks.
-                    self.latest_gps_sample = self.last_good_gps_sample
+            # Read robot object from shared memory.
+            # Object is sent by main process.
+            with self.main_to_remote_lock:
+                recieved_robot_object = pickle.loads(self.main_to_remote_string["value"])
+            time2 = time.time() - debug_time
+        except Exception as e:
+            self.logger.error("Exception reading remote string.")
+            raise(e)
+
+        # TODO: switch to isinstance() without introducing circular dependency.
+        if str(type(recieved_robot_object)) != "<class '__main__.Robot'>":
+            self.logger.info("Waiting for valid robot object before running remote control code.")
+            time.sleep(_SLOW_POLLING_SLEEP_S)
+            return
+
+        self.robot_object = recieved_robot_object
+        self.logger.debug("Remote received new robot object.")
+        time3 = time.time() - debug_time
+        if not self.nav_path.points or self.loaded_path_name != self.robot_object.loaded_path_name:
+            if self.robot_object.loaded_path and self.gps.last_sample() is not None:
+                self.reload_path()
+        self.activate_autonomy = self.robot_object.activate_autonomy
+        # self.autonomy_velocity = self.robot_object.autonomy_velocity
+        # Autonomy is disabled until robot is ready or if joystick is used.
+        if self.autonomy_hold:
+            self.activate_autonomy = False
+        # Reset disabled autonomy if autonomy is turned off in the command.
+        if self.robot_object.clear_autonomy_hold:
+            self.logger.debug("clearing autonomy hold.")
+            self.autonomy_hold = False
+            # Reset disengagement timer.
+            self.disengagement_time = time.time() - _DISENGAGEMENT_RETRY_DELAY_SEC
+
+        time4 = time.time() - debug_time
+
+        #    print("begin robot calc")
+
+        debug_points = (None, None, None, None)
+        gps_path_angle_error = 0
+        absolute_path_distance = math.inf
+        time5 = 0
+        sample = self.gps.last_sample()
+        if self.robot_object is not None and sample is not None:
+            vehicle_front = gps_tools.project_point(sample, sample.azimuth_degrees, 1.0)
+            vehicle_rear = gps_tools.project_point(sample, sample.azimuth_degrees, -1.0)
+
+            time5 = time.time() - debug_time
+            (projected_path_tangent_point, closest_path_point,
+             absolute_path_distance, drive_reverse, calculated_rotation, calculated_strafe) = self.steering_calc()
+
+            # These extra points can be displayed in the web UI.
```

Comment:
The code diff becomes messier from now on. But mostly are just extracting a few methods to make the loop shorter, hence more readable.